org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -306,14 +313,92 @@ public static String getPackageName(IJavaProject javaProject, String fileContent
 		return (pkg == null || pkg.getName() == null)?"":pkg.getName().getFullyQualifiedName();
 	}
 
+	/**
+	 * Returns with the human readable name of the element. For types with type
+	 * arguments, it is {@code Comparable<T>} instead of {@code Comparable}. First,
+	 * this method tries to retrieve the
+	 * {@link JavaElementLabels#getElementLabel(IJavaElement, long) label} of the
+	 * element, then falls back to {@link IJavaElement#getElementName() element
+	 * name}. Returns {@code null} if the argument does not have a name.
+	 */
+	public static String getName(IJavaElement element) {
+		Assert.isNotNull(element, "element");
+		String name = JavaElementLabels.getElementLabel(element, ALL_DEFAULT | M_APP_RETURNTYPE | ROOT_VARIABLE);
+		return name == null ? element.getElementName() : name;
+	}
 
 	/**
-	 * Given the uri returns a {@link IClassFile}.
-	 * May return null if it can not resolve the uri to a
-	 * library.
+	 * Returns with the document symbol {@code SymbolKind kind} for the Java
+	 * element.
+	 */
+	public static SymbolKind getSymbolKind(IJavaElement element) {
+		switch (element.getElementType()) {
+			case IJavaElement.ANNOTATION:
+				return SymbolKind.Property; // TODO: find a better mapping
+			case IJavaElement.CLASS_FILE:
+			case IJavaElement.COMPILATION_UNIT:
+				return SymbolKind.File;
+			case IJavaElement.FIELD:
+				return SymbolKind.Field;
+			case IJavaElement.IMPORT_CONTAINER:
+			case IJavaElement.IMPORT_DECLARATION:
+				return SymbolKind.Module;
+			case IJavaElement.INITIALIZER:
+				return SymbolKind.Constructor;
+			case IJavaElement.LOCAL_VARIABLE:
+			case IJavaElement.TYPE_PARAMETER:
+				return SymbolKind.Variable;
+			case IJavaElement.METHOD:
+				try {
+					// TODO handle `IInitializer`. What should be the `SymbolKind`?

```

Comment:
that's handled L346

Reply:
This method is duplicated with DocumentSymbolHandler#mapKind method, i would remove this one.

Reply:
Have a try on the constructor, the element type for constructor is actually IJavaElement.METHOD. Looks like this logic is required. Given this method is duplicated with DocumentSymbolHandlder#mapKind, i will keep the latter.

For example, hover on the constructor PetClinicApplication() to show call hierarchy, it's type is method. 
```java
	public PetClinicApplication() {
		foo("test", 1);
	}
```